From 7ffef6b52fbbbf250abc4e7ead6362c23379ffa0 Mon Sep 17 00:00:00 2001
From: orenbh <orenbh@marvell.com>
Date: Mon, 14 Nov 2016 13:37:01 +0200
Subject: [PATCH 0617/1345] dts: ap806: add support for CPU Idle power state

commit  9bc883aaa5a90a46aed85108e2adb79380afbe47 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

CPUs currently support only WFI and CPU deep Idle
CPU Idle is enabled by default in kernel defconfig,
however, it will NOT take effect until defined in device-tree.

This patch adds CPU deep Idle and Cluster deep Idle states
BUT it does not define the idle state for each cpu (defined
under cpu-idle-states parameter) therefore it does NOT activate
CPU Idle capability.

In order to activate CPU idle following configuration should be
add under cpuX node (defined under 'cpus' node in AP806
dual/quad dtsi):
cpu-idle-states = <&CPU_SLEEP_0>;

Change-Id: Ibd229707f4de31653a14bb3aa389e8c87cbac1c7
Signed-off-by: orenbh <orenbh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/33743
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-by: Neta Zur Hershkovits <neta@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm64/boot/dts/marvell/armada-ap806-dual.dtsi |    6 +++--
 arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi |    4 +++
 arch/arm64/boot/dts/marvell/armada-ap806.dtsi      |   26 ++++++++++++++++++++
 3 files changed, 34 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/armada-ap806-dual.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806-dual.dtsi
index 25b95cf..37624f3 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806-dual.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806-dual.dtsi
@@ -54,21 +54,23 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		cpu@000 {
+		cpu0: cpu@000 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x000>;
 			clocks = <&ap806_clock 0>;
 			operating-points-v2 = <&cluster0_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
-		cpu@001 {
+		cpu1: cpu@001 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x001>;
 			clocks = <&ap806_clock 0>;
 			operating-points-v2 = <&cluster0_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
 	};
 };
diff --git a/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
index 75f4fdb..ed2822e 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
@@ -61,6 +61,7 @@
 			clocks = <&ap806_clock 0>;
 			operating-points-v2 = <&cluster0_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
 		cpu1: cpu@001 {
 			device_type = "cpu";
@@ -69,6 +70,7 @@
 			clocks = <&ap806_clock 0>;
 			operating-points-v2 = <&cluster0_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
 		cpu2: cpu@100 {
 			device_type = "cpu";
@@ -77,6 +79,7 @@
 			clocks = <&ap806_clock 1>;
 			operating-points-v2 = <&cluster1_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
 		cpu3: cpu@101 {
 			device_type = "cpu";
@@ -85,6 +88,7 @@
 			clocks = <&ap806_clock 1>;
 			operating-points-v2 = <&cluster1_opp>;
 			enable-method = "psci";
+			cpu-idle-states;
 		};
 	};
 
diff --git a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
index 7bd61b0..76dad51 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806.dtsi
@@ -86,6 +86,32 @@
 			clock-latency-ns = <50000>;
 		};
 	};
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		idle_states {
+			entry_method = "arm,pcsi";
+
+			CPU_SLEEP_0: cpu-sleep-0 {
+				compatible = "arm,idle-state";
+				local-timer-stop;
+				arm,psci-suspend-param = <0x0010000>;
+				entry-latency-us = <80>;
+				exit-latency-us  = <160>;
+				min-residency-us = <320>;
+			};
+
+			CLUSTER_SLEEP_0: cluster-sleep-0 {
+				compatible = "arm,idle-state";
+				local-timer-stop;
+				arm,psci-suspend-param = <0x1010000>;
+				entry-latency-us = <500>;
+				exit-latency-us = <1000>;
+				min-residency-us = <2500>;
+			};
+		};
+	};
 
 	ap806 {
 		#address-cells = <2>;
-- 
1.7.9.5

